Pixhawk 6X-RT Add bootloader update guide

PetervdPerk-NXP opened the pull request:
Add guide howto update bootloader for V6X-RT with SWD and ISP through usb in case when bricked.

hamishwillee commented on en/advanced_config/bootloader_update_v6xrt.md:
```diff
@@ -0,0 +1,163 @@
+1. Power on the Pixhawk with another USB cable and connect the Dronecode probe to the FMU-DEBUG port.
+
+   :::note
+   To be able to connect the Dronecode probe to the FMU-DEBUG port
```
This sentence appears truncated?

hamishwillee commented:
Thanks so much @PetervdPerk-NXP .

I've subedited a bit - basically just put the steps in bullets and a few typos. Also moved images into a subfolder.

The logical place for this in the tree appears to be below https://docs.px4.io/main/en/advanced_config/bootloader_update.html so I put it as a peer of that document.

We should "differentiate" this from https://docs.px4.io/main/en/advanced_config/bootloader_update.html - I mean "why do we need a special set of bootloader instruction"? If there is material that should be in one or the other of these docs and cross linked?
Minimally the https://docs.px4.io/main/en/advanced_config/bootloader_update.html should have a line stating that for RT use these instructions instead - can you add it?

PetervdPerk-NXP commented:
So the "Dronecode Probe Bootloader Update" procedure mostly the same.

But in case people don't have a debug probe the imxrt mcu has a build-in "first-stage" bootloader that can update the "second-stage" px4 bootloader. Also the "first-stage" bootloader can be started using the micro switch in case of a software brick.

Overall I think the "Dronecode Probe Bootloader Update" could use some instruction updates and in the case of FMUv2 and Fmuv6x-RT there a special extra instructions to do it without a debug probe. As far as I know the other FMUvX require a probe to recover/update the bootloader. @david_s5 am I correct?

hamishwillee commented:
> Overall I think the "Dronecode Probe Bootloader Update" could use some instruction updates. Ideally removing the arm-none-eabi-gdb way of flashing entirely and add specific instructions using for isntance JLinkExe and/or pyocd

Can you do this?

For now, can we merge the dronecode section into https://docs.px4.io/main/en/advanced_config/bootloader_update.html#dronecode-probe-bootloader-update in some way, then cross link from the new doc? (i.e. link rather than reproduce)? Then this doc can be just the "special magic for 6xrt"
 
Then perhaps add a specific note in that doc in the introduction "If you want to bootload on 6xrt then see this doc instead".

PetervdPerk-NXP commented:
Sure I could try to do a generic guide for flashing the bootloader using the dronecode probe and the remove the chapter from the from v6xrt specific instructions.

hamishwillee commented:
> We might want to change this then as well
> https://docs.px4.io/main/en/config/firmware.html#bootloader-update
> And forward to the advanced - > bootloader section

We do already - see the end of that very short section. We could cut it back even further to just the link, as you suggest.

hamishwillee commented:
> Sure I could try to do a generic guide for flashing the bootloader using the dronecode probe and the remove the chapter from the from v6xrt specific instructions.

Thanks! I'll assume you will do it in this PR. But if you want to do it as a post-process let me know.

davids5 commented:
> So the "Dronecode Probe Bootloader Update" procedure mostly the same.
> 
> But in case people don't have a debug probe the imxrt mcu has a build-in "first-stage" bootloader that can update the "second-stage" px4 bootloader. Also the "first-stage" bootloader can be started using the micro switch in case of a software brick.
> 
> Overall I think the "Dronecode Probe Bootloader Update" could use some instruction updates. Ideally removing the `arm-none-eabi-gdb` way of flashing entirely and add specific instructions using for isntance `JLinkExe` and/or `pyocd`
> 
> In the case of FMUv2 and Fmuv6x-RT there are special extra instructions to do it without a debug probe. As far as I know the other FMUvX require a probe to recover/update the bootloader. @davids5 am I correct?

If not bricked. There is a bl_update applications that CAN be used to update a bootloader.

davids5 commented on en/advanced_config/bootloader_update_v6xrt.md:
````diff
@@ -0,0 +1,163 @@
+1. Go into the directory containing the binary and run the following command in the terminal:
+
+   ```sh
+   arm-none-eabi-gdb px4_fmu-v6xrt_default.elf
````
bootloader?

PetervdPerk-NXP commented:
@hamishwillee 
In the bootloader_update section I've renamed "Dronecode Probe Bootloader Update" to "Debug Probe Bootloader Update" and made generic for every board. I've added a step with a hyperlink to `swd_debug.md#debug-probes-for-px4-hardware` so people can use their favorite debug probe.

Also added a chapter for the MCU-Link debug probe.

davids5 commented on en/debug/probe_mculink.md:
```diff
@@ -0,0 +1,78 @@
+# MCU-Link Debug Probe
+
+The [MCU-Link Debug Probe](https://www.nxp.com/design/design-center/software/development-software/mcuxpresso-software-and-tools-/mcu-link-debug-probe:MCU-LINK) is a cheap, fast and highly capable debug probe that can serve as a stand-alone debug and console communicator for a PX4 developer:
```
@PetervdPerk-NXP Have you verified the operation on all the targets?

PetervdPerk-NXP replied:
Tested on pixhawk 5x and 6c

hamishwillee replied:
6x?

PetervdPerk-NXP replied:
I don't have a 6X but let's assume it's the same STM32H7 it should just work

davids5 commented on en/advanced_config/bootloader_update.md:
```diff
@@ -96,18 +98,24 @@ To update the bootloader:
    :::
 
-## Dronecode Probe Bootloader Update
+## Debug Probe Bootloader Update
```
@mrpollo  Is this correct and OK to change

hamishwillee replied:
This is _good_ to change, provided the instructions worth with all the linked probes. I will assume so.

hamishwillee commented on en/flight_controller/pixhawk_series.md:
```diff
@@ -87,8 +87,14 @@ At very high level, the main differences are:
-  .([Pixhawk 5X](../flight_controller/pixhawk5x.md), Skynode)
+  Much more configurable & customizable.
+  ([Pixhawk 5X](../flight_controller/pixhawk5x.md), Skynode)
+- **FMUv6C:** 
+  ([Holybro Pixhawk 6C Mini](../flight_controller/pixhawk6c_mini.md), [Holybro Pixhawk 6C](../flight_controller/pixhawk6c.md))
+- **FMUv6X:** 
+  ([CUAV Pixhawk V6X](../flight_controller/cuav_pixhawk_v6x.md),[Holybro Pixhawk 6X](../flight_controller/pixhawk6x.md))
+- **FMUv6X-RT:**
+  ([Holybro Pixhawk 6X-RT](../flight_controller/pixhawk6x-rt.md))
```
I noticed these boards are missing a "relative overview". While we're here, can you suggest some text?

PetervdPerk-NXP replied:
Good one hard to say to be honest, V6X/V6X-RT has the same PAB as V5X.
I think but not sure that both V6X/V6X-RT support CAN FD instead of CAN.
Faster MCU core (480Mhz V6X, 1Ghz V6X-RT) More RAM (1MB on V6X, 2MB on V6X-RT), V6X-RT has more flash compared to V6X/V5X 64Mb instead of 2Mb.

hamishwillee replied:
Thanks. I added this. Will try get more info for the others.

hamishwillee commented on en/advanced_config/bootloader_update.md:
````diff
@@ -96,18 +97,25 @@ To update the bootloader:
+
 1. Go into the directory containing the binary and run the following command in the terminal:
 
    ```sh
    arm-none-eabi-gdb px4fmuv5_bl.elf
    ```
 
+   :::note
+   If the hardware is FMUv6X or newer use the in-tree PX4 bootloader.
+   :::
````
I think this would be better? (please check command)

```suggestion
   ```sh
   arm-none-eabi-gdb px4fmuv6x_bl.elf
   ```

   :::note
   The naming pattern for the bootloader binary is the same for both in-tree and legacy PX4 bootloader.
   :::
```

hamishwillee commented on en/advanced_config/bootloader_update.md:
````diff
@@ -96,18 +97,25 @@ To update the bootloader:
 1. Go into the directory containing the binary and run the following command in the terminal:
 
    ```sh
    arm-none-eabi-gdb px4fmuv5_bl.elf
````
```suggestion
   arm-none-eabi-gdb px4fmuv6x_bl.elf
```

PetervdPerk-NXP replied:
This is incorrect so legacy for V5 is 
`px4fmuv5_bl.elf`
Older version I don't even know.
For V6X it's
`px4_fmu-v6x_bootloader.elf`
For V6X-RT it's
`px4_fmu-v6xrt_bootloader.elf`

davids5 replied:
In-tree H7 Booloaders from https://github.com/PX4/PX4-Autopilot are named *._bootloader all the others are from https://github.com/PX4/PX4-Bootloader and end with _bl. 

Using up a term like legacy will not help anyone nor will it be valid on the next bootloaderr.

hamishwillee replied:
Thanks both. Fixed by listing the ones above and adding a note about the pattern.

> Using up a term like legacy will not help anyone nor will it be valid on the next bootloaderr.

You could not be more right. Sorry about that.

hamishwillee commented on en/advanced_config/bootloader_update.md:
````diff
@@ -96,18 +97,25 @@ To update the bootloader:
    ```
 
+   :::note
+   If the hardware is FMUv6X or newer use the in-tree PX4 bootloader.
````
```suggestion
   The naming pattern for the bootloader binary is the same for both in-tree and legacy PX4 bootloader.
```

PetervdPerk-NXP replied:
In correct see comment above.

hamishwillee commented on en/advanced_config/bootloader_update.md:
````diff
@@ -128,20 +136,21 @@ The following steps explain how you can "manually" update the bootloader using t
    ```
 
-1. Power on the Pixhawk with another USB cable and connect the Dronecode probe to the FMU-DEBUG port.
+1. Power on the Pixhawk with another USB cable and connect the probe to the `FMU-DEBUG` port.
````
Assumption here is all probes connect to this port.

PetervdPerk-NXP replied:
Yes correct

hamishwillee commented on en/advanced_config/bootloader_update.md:
```diff
@@ -128,20 +136,21 @@ The following steps explain how you can "manually" update the bootloader using t
 
    :::note
-   To be able to connect the Dronecode probe to the FMU-DEBUG port, you may need to remove the case (e.g. on Pixhawk 4 you would do this using a T6 Torx screwdriver).
+   To be able to connect a Dronecode probe to the `FMU-DEBUG` port, you may need to remove the case (e.g. on Pixhawk 4 you would do this using a T6 Torx screwdriver).
```
```suggestion
   If using a Dronecode probe you may need to remove the case in order to connect to the `FMU-DEBUG` port (e.g. on Pixhawk 4 you would do this using a T6 Torx screwdriver).
```

hamishwillee commented on en/debug/probe_mculink.md:
```diff
@@ -0,0 +1,83 @@
+The [Pixhawk Debug Adapter](https://holybro.com/products/pixhawk-debug-adapter) provides an easy way to connect a Pixhawk to an MCU-Link (the probe does not come with an adapter for working with Pixhawk flight controllers).
+
+:::note
+These instructions have been tested on: FMUv6X-RT, FMUv6X, 6c, FMUv5X.
```
```suggestion
These instructions have been tested on: FMUv6X-RT, FMUv6X, FMUv6c, FMUv5X.
```

hamishwillee commented on en/debug/probe_mculink.md:
```diff
@@ -0,0 +1,83 @@
+The [Pixhawk Debug Adapter](https://holybro.com/products/pixhawk-debug-adapter) provides an easy way to connect a Pixhawk to an MCU-Link (the probe does not come with an adapter for working with Pixhawk flight controllers).
+
+:::note
+These instructions have been tested on: FMUv6X-RT, FMUv6X, FMUv6c, FMUv5X.
```
I presume it works on these boards based on your comment (you didn't mention 6x, but that is mentioned lower down)

hamishwillee commented on en/debug/probe_mculink.md:
```diff
@@ -0,0 +1,83 @@
+The target needs to be one of:
+
+- FMUv6X-RT: `mimxrt1170_cm7`
+- FMUv6X: `stm32h743xx`
```
6C as well?

hamishwillee commented:
@PetervdPerk-NXP Thanks very much for these changes - I like the structural changes. Have added a few minor requests for clarification.

I did some minor structural changes. Specifically I've made the USB version more clearly a sub-topic - and spelled out your point that it is an option if you don't have a debug probe. I also pushed the debug probe option above the QGC update option. Reason being that while the QGC option is the easiest, the Debug option is what most people on most boards will actually have to use.

hamishwillee commented:
Thanks very much @PetervdPerk-NXP (and @davids5). Merging. We can iterate if needed (though I am away for a few weeks)
